Index: openwrt/target/linux/kirkwood/image/Makefile
===================================================================
--- openwrt/target/linux/kirkwood/image/Makefile	(revision 29704)
+++ openwrt/target/linux/kirkwood/image/Makefile	(working copy)
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2009-2010 OpenWrt.org
+# Copyright (C) 2009-2011 OpenWrt.org
 #
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
@@ -8,9 +8,11 @@
 include $(INCLUDE_DIR)/image.mk
 
 JFFS2OPTS += --little-endian --pagesize=0x800 --no-cleanmarkers --pad
+JFFS2_BLOCKSIZE = 128k
 
 define Image/Prepare
 	cp $(LINUX_DIR)/arch/arm/boot/uImage $(KDIR)/uImage
+	cp $(LINUX_DIR)/arch/arm/boot/zImage $(KDIR)/zImage
 endef
 
 define Image/BuildKernel
@@ -31,6 +33,10 @@
 endef
 
 define Image/Build/squashfs
+	$(STAGING_DIR_HOST)/bin/mkimage -A arm -O linux -T multi \
+	-C none -a 0x00008000 -e 0x00008000 -n 'Linux-$(LINUX_VERSION)-initrd' \
+	-d $(KDIR)/zImage:$(KDIR)/root.squashfs $(BIN_DIR)/$(IMG_PREFIX)-$(PROFILE)-$(1).uimg
+
 	$(call prepare_generic_squashfs,$(KDIR)/root.squashfs)
 	( \
 		dd if=$(KDIR)/uImage bs=4096k conv=sync; \
Index: openwrt/target/linux/kirkwood/profiles/200-Dockstar.mk
===================================================================
--- openwrt/target/linux/kirkwood/profiles/200-Dockstar.mk	(revision 29704)
+++ openwrt/target/linux/kirkwood/profiles/200-Dockstar.mk	(working copy)
@@ -6,12 +6,12 @@
 #
 
 define Profile/Dockstar
-  NAME:=Seagate Dockstar
-  PACKAGES:=
+	NAME:=Seagate Dockstar
+	PACKAGES:=
 endef
 
 define Profile/Dockstar/Description
-        Seagate Dockstar Profile
+	Seagate Dockstar Profile
 endef
 
 $(eval $(call Profile,Dockstar))
Index: openwrt/target/linux/kirkwood/profiles/400-rd88f6281.mk
===================================================================
--- openwrt/target/linux/kirkwood/profiles/400-rd88f6281.mk	(revision 0)
+++ openwrt/target/linux/kirkwood/profiles/400-rd88f6281.mk	(revision 0)
@@ -0,0 +1,17 @@
+#
+# Copyright (C) 2011 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/rd88f6281
+	NAME:=Marvell RD88F6281 Reference Boards
+	PACKAGES:=mtd-utils kmod-ath9k kmod-mwl8k wpad-mini
+endef
+
+define Profile/rd88f6281/Description
+	Marvell RD88F6281 Reference Boards Profile
+endef
+
+$(eval $(call Profile,rd88f6281))
Index: openwrt/target/linux/kirkwood/profiles/100-Sheevaplug.mk
===================================================================
--- openwrt/target/linux/kirkwood/profiles/100-Sheevaplug.mk	(revision 29704)
+++ openwrt/target/linux/kirkwood/profiles/100-Sheevaplug.mk	(working copy)
@@ -6,12 +6,12 @@
 #
 
 define Profile/Sheevaplug
-  NAME:=Globalscale Sheevaplug
-  PACKAGES:=
+	NAME:=Globalscale Sheevaplug
+	PACKAGES:=
 endef
 
 define Profile/Sheevaplug/Description
-        Globalscale Sheevaplug Profile
+	Globalscale Sheevaplug Profile
 endef
 
 $(eval $(call Profile,Sheevaplug))
Index: openwrt/target/linux/kirkwood/profiles/300-Iconnect.mk
===================================================================
--- openwrt/target/linux/kirkwood/profiles/300-Iconnect.mk	(revision 29704)
+++ openwrt/target/linux/kirkwood/profiles/300-Iconnect.mk	(working copy)
@@ -6,12 +6,12 @@
 #
 
 define Profile/Iconnect
-  NAME:=Iomega iConnect Wireless
-  PACKAGES:=kmod-i2c-mv64xxx kmod-hwmon-core kmod-hwmon-lm63
+	NAME:=Iomega iConnect Wireless
+	PACKAGES:=kmod-hwmon-core kmod-hwmon-lm63
 endef
 
 define Profile/Iconnect/Description
-        Iomega iConnect Wireless
+	Iomega iConnect Wireless
 endef
 
 $(eval $(call Profile,Iconnect))
Index: openwrt/target/linux/kirkwood/base-files-rd88f6281/etc/config/network
===================================================================
--- openwrt/target/linux/kirkwood/base-files-rd88f6281/etc/config/network	(revision 0)
+++ openwrt/target/linux/kirkwood/base-files-rd88f6281/etc/config/network	(revision 0)
@@ -0,0 +1,25 @@
+# Copyright (C) 2011 OpenWrt.org
+
+config 'interface' 'loopback'
+	option 'ifname' 'lo'
+	option 'proto' 'static'
+	option 'ipaddr' '127.0.0.1'
+	option 'netmask' '255.0.0.0'
+
+config 'interface' 'eth0'
+	option 'ifname' 'eth0'
+	option 'proto' 'none'
+
+config 'interface' 'wan'
+	option 'hostname' 'OpenWrt'
+	option 'proto' 'dhcp'
+	option 'type' 'bridge'
+	option 'ifname' 'wan eth1 eth2'
+
+config 'interface' 'lan'
+	option 'type' 'bridge'
+	option 'proto' 'static'
+	option 'netmask' '255.255.255.0'
+	option 'nat' '1'
+	option 'ipaddr' '192.168.2.1'
+	option 'ifname' 'lan1 lan2 lan3 lan4'
Index: openwrt/target/linux/kirkwood/patches/200-iconnect.patch
===================================================================
--- openwrt/target/linux/kirkwood/patches/200-iconnect.patch	(revision 29704)
+++ openwrt/target/linux/kirkwood/patches/200-iconnect.patch	(working copy)
@@ -35,7 +35,7 @@
  #define MPP12_SPI_MOSI		MPP( 12, 0xb, 0, 1, 0,   0,   0,   0,   1 )
 --- /dev/null
 +++ b/arch/arm/mach-kirkwood/iconnect-setup.c
-@@ -0,0 +1,188 @@
+@@ -0,0 +1,203 @@
 +/*
 + * arch/arm/mach-kirkwood/iconnect-setup.c
 + *
@@ -57,6 +57,7 @@
 +#include <linux/gpio_keys.h>
 +#include <linux/input.h>
 +#include <linux/leds.h>
++#include <linux/i2c.h>
 +#include <asm/mach-types.h>
 +#include <asm/mach/arch.h>
 +#include <mach/kirkwood.h>
@@ -89,6 +90,11 @@
 +
 +static struct gpio_led iconnect_led_pins[] = {
 +	{
++		.name			= "iconnect:led_level",
++		.default_trigger	= "default-on",
++		.gpio			= 41,
++	},
++	{
 +		.name			= "iconnect:blue:power",
 +		.default_trigger	= "default-on",
 +		.gpio			= 42,
@@ -174,6 +180,7 @@
 +static unsigned int iconnect_mpp_config[] __initdata = {
 +	MPP12_GPIO, /*Input for reset button*/
 +	MPP35_GPIO, /*Input for OTB button*/
++	MPP41_GPIO,
 +	MPP42_GPIO,
 +	MPP43_GPIO,
 +	MPP44_GPIO,
@@ -184,6 +191,10 @@
 +	0
 +};
 +
++static struct i2c_board_info __initdata iconnect_i2c_rtc = {
++	I2C_BOARD_INFO("lm63", 0x4c),
++};
++
 +static void __init iconnect_init(void)
 +{
 +	u32 dev, rev;
@@ -201,10 +212,13 @@
 +	kirkwood_pcie_id(&dev, &rev);
 +
 +	kirkwood_uart0_init();
-+	kirkwood_i2c_init();
 +
 +	platform_device_register(&iconnect_leds);
 +	platform_device_register(&iconnect_button_device);
++	
++	kirkwood_i2c_init();
++	i2c_register_board_info(0, &iconnect_i2c_rtc, 1);
++
 +}
 +
 +static int __init iconnect_pci_init(void)
@@ -221,6 +235,7 @@
 +	.boot_params	= 0x00000100,
 +	.init_machine	= iconnect_init,
 +	.map_io		= kirkwood_map_io,
++	.init_early	= kirkwood_init_early,
 +	.init_irq	= kirkwood_init_irq,
 +	.timer		= &kirkwood_timer,
 +MACHINE_END
Index: openwrt/target/linux/kirkwood/patches/001-openwrt_partition_map.patch
===================================================================
--- openwrt/target/linux/kirkwood/patches/001-openwrt_partition_map.patch	(revision 29704)
+++ openwrt/target/linux/kirkwood/patches/001-openwrt_partition_map.patch	(working copy)
@@ -1,3 +1,14 @@
+--- a/arch/arm/mach-kirkwood/rd88f6281-setup.c
++++ b/arch/arm/mach-kirkwood/rd88f6281-setup.c
+@@ -34,7 +34,7 @@ static struct mtd_partition rd88f6281_na
+ 		.offset = MTDPART_OFS_NXTBLK,
+ 		.size = SZ_2M
+ 	}, {
+-		.name = "root",
++		.name = "rootfs",
+ 		.offset = MTDPART_OFS_NXTBLK,
+ 		.size = MTDPART_SIZ_FULL
+ 	},
 --- a/arch/arm/mach-kirkwood/sheevaplug-setup.c
 +++ b/arch/arm/mach-kirkwood/sheevaplug-setup.c
 @@ -33,7 +33,7 @@ static struct mtd_partition sheevaplug_n
Index: openwrt/target/linux/kirkwood/patches/300-mv88f6281_rev_z0_tclock.patch
===================================================================
--- openwrt/target/linux/kirkwood/patches/300-mv88f6281_rev_z0_tclock.patch	(revision 0)
+++ openwrt/target/linux/kirkwood/patches/300-mv88f6281_rev_z0_tclock.patch	(revision 0)
@@ -0,0 +1,11 @@
+--- a/arch/arm/mach-kirkwood/common.c
++++ b/arch/arm/mach-kirkwood/common.c
+@@ -854,7 +854,7 @@ int __init kirkwood_find_tclk(void)
+ 
+ 	kirkwood_pcie_id(&dev, &rev);
+ 
+-	if (dev == MV88F6281_DEV_ID || dev == MV88F6282_DEV_ID)
++	if ((dev == MV88F6281_DEV_ID && rev >= MV88F6281_REV_A0) || dev == MV88F6282_DEV_ID)
+ 		if (((readl(SAMPLE_AT_RESET) >> 21) & 1) == 0)
+ 			return 200000000;
+ 
Index: openwrt/target/linux/kirkwood/config-default
===================================================================
--- openwrt/target/linux/kirkwood/config-default	(revision 29704)
+++ openwrt/target/linux/kirkwood/config-default	(working copy)
@@ -1,21 +1,26 @@
 CONFIG_ALIGNMENT_TRAP=y
+CONFIG_ARCH_HAS_CPU_IDLE_WAIT=y
 CONFIG_ARCH_KIRKWOOD=y
-# CONFIG_ARCH_NUC93X is not set
 CONFIG_ARCH_REQUIRE_GPIOLIB=y
 # CONFIG_ARCH_SELECT_MEMORY_MODEL is not set
 # CONFIG_ARCH_SPARSEMEM_DEFAULT is not set
 # CONFIG_ARCH_SUPPORTS_MSI is not set
-CONFIG_ARCH_SUSPEND_POSSIBLE=y
 # CONFIG_ARCH_USES_GETTIMEOFFSET is not set
 CONFIG_ARM=y
 CONFIG_ARM_L1_CACHE_SHIFT=5
 # CONFIG_ARM_THUMB is not set
 # CONFIG_ARPD is not set
-CONFIG_BITREVERSE=y
+CONFIG_ATA=y
+CONFIG_BCMA_POSSIBLE=y
+CONFIG_BLK_DEV_RAM=y
+CONFIG_BLK_DEV_RAM_COUNT=16
+CONFIG_BLK_DEV_RAM_SIZE=65536
 CONFIG_BLK_DEV_SD=y
 CONFIG_CACHE_FEROCEON_L2=y
 # CONFIG_CACHE_FEROCEON_L2_WRITETHROUGH is not set
-CONFIG_CMDLINE="rootdelay=1 root=/dev/mmcblk0p1 noinitrd console=ttyS0,115200"
+CONFIG_CLKSRC_MMIO=y
+CONFIG_CMDLINE="rootdelay=1 console=ttyS0,115200"
+CONFIG_CMDLINE_FROM_BOOTLOADER=y
 CONFIG_CPU_32v5=y
 CONFIG_CPU_ABRT_EV5T=y
 CONFIG_CPU_CACHE_VIVT=y
@@ -29,24 +34,35 @@
 CONFIG_CPU_IDLE_GOV_LADDER=y
 CONFIG_CPU_PABRT_LEGACY=y
 CONFIG_CPU_TLB_FEROCEON=y
-CONFIG_DEBUG_KERNEL=y
+CONFIG_CPU_USE_DOMAINS=y
+CONFIG_CRC16=y
+CONFIG_CRYPTO_AES=y
+CONFIG_CRYPTO_ALGAPI=y
+CONFIG_CRYPTO_ALGAPI2=y
+CONFIG_CRYPTO_BLKCIPHER2=y
+CONFIG_CRYPTO_DEFLATE=y
+CONFIG_CRYPTO_DEV_MV_CESA=y
+CONFIG_CRYPTO_HASH2=y
+CONFIG_CRYPTO_HW=y
+CONFIG_CRYPTO_LZO=y
+CONFIG_CRYPTO_RNG2=y
+CONFIG_CRYPTO_WORKQUEUE=y
+CONFIG_DEBUG_BUGVERBOSE=y
 # CONFIG_DEBUG_USER is not set
 CONFIG_DECOMPRESS_LZMA=y
-CONFIG_DEVPORT=y
-# CONFIG_DLCI is not set
-# CONFIG_DM9000 is not set
 CONFIG_DNOTIFY=y
+CONFIG_EFI_PARTITION=y
 CONFIG_EXT4_FS=y
-# CONFIG_FPE_FASTFPE is not set
-# CONFIG_FPE_NWFPE is not set
+CONFIG_EXT4_FS_XATTR=y
 CONFIG_FRAME_POINTER=y
 CONFIG_GENERIC_ATOMIC64=y
 CONFIG_GENERIC_CLOCKEVENTS=y
 CONFIG_GENERIC_CLOCKEVENTS_BUILD=y
-CONFIG_GENERIC_FIND_LAST_BIT=y
 CONFIG_GENERIC_GPIO=y
-CONFIG_GENERIC_HARDIRQS_NO__DO_IRQ=y
+CONFIG_GENERIC_IRQ_CHIP=y
+CONFIG_GENERIC_IRQ_SHOW=y
 CONFIG_GPIOLIB=y
+CONFIG_GPIO_SYSFS=y
 # CONFIG_HAMRADIO is not set
 CONFIG_HARDIRQS_SW_RESEND=y
 CONFIG_HAS_DMA=y
@@ -54,29 +70,45 @@
 CONFIG_HAS_IOPORT=y
 CONFIG_HAVE_AOUT=y
 CONFIG_HAVE_ARCH_KGDB=y
+CONFIG_HAVE_ARCH_PFN_VALID=y
+CONFIG_HAVE_C_RECORDMCOUNT=y
+CONFIG_HAVE_DMA_API_DEBUG=y
+CONFIG_HAVE_DYNAMIC_FTRACE=y
+CONFIG_HAVE_FTRACE_MCOUNT_RECORD=y
+CONFIG_HAVE_FUNCTION_GRAPH_TRACER=y
 CONFIG_HAVE_FUNCTION_TRACER=y
 CONFIG_HAVE_GENERIC_DMA_COHERENT=y
+CONFIG_HAVE_GENERIC_HARDIRQS=y
 CONFIG_HAVE_IDE=y
+CONFIG_HAVE_IRQ_WORK=y
 CONFIG_HAVE_KERNEL_GZIP=y
 CONFIG_HAVE_KERNEL_LZMA=y
 CONFIG_HAVE_KERNEL_LZO=y
-CONFIG_HAVE_KPROBES=y
-CONFIG_HAVE_KRETPROBES=y
+CONFIG_HAVE_KERNEL_XZ=y
 CONFIG_HAVE_LATENCYTOP_SUPPORT=y
+CONFIG_HAVE_MEMBLOCK=y
 CONFIG_HAVE_OPROFILE=y
 CONFIG_HAVE_PERF_EVENTS=y
 CONFIG_HAVE_PROC_CPU=y
+CONFIG_HAVE_REGS_AND_STACK_ACCESS_API=y
+CONFIG_HAVE_SCHED_CLOCK=y
+CONFIG_HAVE_SPARSE_IRQ=y
 CONFIG_HW_RANDOM=y
-# CONFIG_I2C_MV64XXX is not set
+CONFIG_I2C=y
+CONFIG_I2C_BOARDINFO=y
+CONFIG_I2C_CHARDEV=y
+CONFIG_I2C_MV64XXX=y
+# CONFIG_I2C_PXA_PCI is not set
 CONFIG_INET_LRO=y
 CONFIG_INITRAMFS_SOURCE=""
-# CONFIG_IP_ADVANCED_ROUTER is not set
-CONFIG_IP_PIMSM_V1=y
-CONFIG_IP_PIMSM_V2=y
-CONFIG_JBD=y
+# CONFIG_ISDN is not set
+CONFIG_JBD2=y
+CONFIG_KTIME_SCALAR=y
 CONFIG_LEDS_GPIO=y
 CONFIG_LEGACY_PTYS=y
 CONFIG_LEGACY_PTY_COUNT=256
+CONFIG_LZO_COMPRESS=y
+CONFIG_LZO_DECOMPRESS=y
 # CONFIG_MACH_D2NET_V2 is not set
 # CONFIG_MACH_DB88F6281_BP is not set
 CONFIG_MACH_DOCKSTAR=y
@@ -89,15 +121,19 @@
 # CONFIG_MACH_NET5BIG_V2 is not set
 # CONFIG_MACH_NETSPACE_MAX_V2 is not set
 # CONFIG_MACH_NETSPACE_V2 is not set
+CONFIG_MACH_OPENRD=y
 CONFIG_MACH_OPENRD_BASE=y
 CONFIG_MACH_OPENRD_CLIENT=y
 # CONFIG_MACH_OPENRD_ULTIMATE is not set
 # CONFIG_MACH_RD88F6192_NAS is not set
-# CONFIG_MACH_RD88F6281 is not set
+CONFIG_MACH_RD88F6281=y
 CONFIG_MACH_SHEEVAPLUG=y
 # CONFIG_MACH_T5325 is not set
 # CONFIG_MACH_TS219 is not set
 # CONFIG_MACH_TS41X is not set
+CONFIG_MARVELL_PHY=y
+# CONFIG_MFD_SUPPORT is not set
+# CONFIG_MISC_DEVICES is not set
 CONFIG_MMC=y
 CONFIG_MMC_BLOCK=y
 CONFIG_MMC_MVSDIO=y
@@ -110,17 +146,24 @@
 # CONFIG_MTD_ROOTFS_ROOT_DEV is not set
 # CONFIG_MTD_ROOTFS_SPLIT is not set
 # CONFIG_MTD_SM_COMMON is not set
+CONFIG_MTD_UBI=y
+CONFIG_MTD_UBI_BEB_RESERVE=1
+# CONFIG_MTD_UBI_DEBUG is not set
+# CONFIG_MTD_UBI_GLUEBI is not set
+CONFIG_MTD_UBI_WL_THRESHOLD=4096
 CONFIG_MV643XX_ETH=y
 CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_NEED_PER_CPU_KM=y
 CONFIG_NET_DSA=y
 # CONFIG_NET_DSA_MV88E6060 is not set
-# CONFIG_NET_DSA_MV88E6123_61_65 is not set
+CONFIG_NET_DSA_MV88E6123_61_65=y
 CONFIG_NET_DSA_MV88E6131=y
 CONFIG_NET_DSA_MV88E6XXX=y
 CONFIG_NET_DSA_MV88E6XXX_NEED_PPU=y
 CONFIG_NET_DSA_TAG_DSA=y
-# CONFIG_NET_DSA_TAG_EDSA is not set
+CONFIG_NET_DSA_TAG_EDSA=y
 # CONFIG_NET_DSA_TAG_TRAILER is not set
+# CONFIG_NET_ETHERNET is not set
 CONFIG_NLS=y
 CONFIG_OUTER_CACHE=y
 CONFIG_PAGEFLAGS_EXTENDED=y
@@ -129,25 +172,39 @@
 CONFIG_PERF_USE_VMALLOC=y
 CONFIG_PHYLIB=y
 CONFIG_PLAT_ORION=y
+# CONFIG_PREEMPT_RCU is not set
+# CONFIG_QUOTACTL is not set
 CONFIG_RTC_CLASS=y
 CONFIG_RTC_DRV_MV=y
+CONFIG_SATA_MV=y
+CONFIG_SATA_PMP=y
 CONFIG_SCSI=y
-CONFIG_SCSI_MOD=y
 # CONFIG_SCSI_MULTI_LUN is not set
 # CONFIG_SDIO_UART is not set
-# CONFIG_SERIAL_8250_EXTENDED is not set
 CONFIG_SPLIT_PTLOCK_CPUS=999999
+# CONFIG_STAGING is not set
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
+CONFIG_UACCESS_WITH_MEMCPY=y
+CONFIG_UBIFS_FS=y
+# CONFIG_UBIFS_FS_ADVANCED_COMPR is not set
+# CONFIG_UBIFS_FS_DEBUG is not set
+CONFIG_UBIFS_FS_LZO=y
+# CONFIG_UBIFS_FS_XATTR is not set
+CONFIG_UBIFS_FS_ZLIB=y
 CONFIG_UID16=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_EHCI_TT_NEWSCHED=y
 CONFIG_USB_STORAGE=y
 CONFIG_USB_SUPPORT=y
 # CONFIG_USB_UHCI_HCD is not set
 CONFIG_VECTORS_BASE=0xffff0000
 # CONFIG_VFP is not set
 CONFIG_VM_EVENT_COUNTERS=y
-CONFIG_WAN=y
+# CONFIG_WATCHDOG is not set
+CONFIG_XZ_DEC=y
+CONFIG_XZ_DEC_ARM=y
+CONFIG_XZ_DEC_BCJ=y
 CONFIG_ZBOOT_ROM_BSS=0x0
 CONFIG_ZBOOT_ROM_TEXT=0x0
 CONFIG_ZONE_DMA_FLAG=0
Index: openwrt/target/linux/kirkwood/base-files/sbin/install2ubi
===================================================================
--- openwrt/target/linux/kirkwood/base-files/sbin/install2ubi	(revision 0)
+++ openwrt/target/linux/kirkwood/base-files/sbin/install2ubi	(revision 0)
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+# Copyright (C) 2011 Marvell Smiconductors
+
+read -p "Do you want to install currently running firmware to NAND flash using UBI, [y/n]?" input
+
+if [ "${input}" == "y" ]; then
+	echo "Performing install into NAND flash using UBI..."
+	ubiformat /dev/mtd2 -s 512 && ubiattach /dev/ubi_ctrl -m 2 && ubimkvol /dev/ubi0 -N rootfs -m && mount -t ubifs ubi0:rootfs /mnt
+	cd /; ls -1 | grep -vE "proc|sys|tmp|mnt" | awk '{print "cp -a "$1" /mnt/"}' | sh; mkdir -p /mnt/proc /mnt/sys /mnt/tmp /mnt/mnt; cd -
+	echo "Performed install into NAND flash using UBI, you may reboot into new system."
+else
+	echo "Skipping install into NAND flash."
+fi
Index: openwrt/target/linux/kirkwood/Makefile
===================================================================
--- openwrt/target/linux/kirkwood/Makefile	(revision 29704)
+++ openwrt/target/linux/kirkwood/Makefile	(working copy)
@@ -9,10 +9,10 @@
 ARCH:=arm
 BOARD:=kirkwood
 BOARDNAME:=Marvell Kirkwood
-FEATURES:=targz usb jffs2
+FEATURES:=targz usb jffs2 squashfs
 MAINTAINER:=Imre Kaloz <kaloz@openwrt.org>
 
-LINUX_VERSION:=2.6.37.6
+LINUX_VERSION:=3.1.7
 
 include $(INCLUDE_DIR)/target.mk
 
Index: openwrt/target/linux/kirkwood/README
===================================================================
--- openwrt/target/linux/kirkwood/README	(revision 0)
+++ openwrt/target/linux/kirkwood/README	(revision 0)
@@ -0,0 +1,37 @@
+This Kirkwood target is intended to be used with devices based on Marvell Kirkwood 6281/2 ARM-compatible CPU (aka Feroceon).
+
+Among supported devices are:
+	* Globalscale Sheevaplug
+	* Seagate Dockstar
+	* Iomega iConnect Wireless
+	* Marvell RD88F6281 Reference Boards
+
+Marvell RD88F6281 Reference Boards currently utilize UBIFS on NAND scheme for rootfs
+and uImage is stored in raw NAND area.
+Currently only manual install/upgrade procedure tested.
+To perform manual/initial installation you will need a TFTP server and board specific images,
+for example: openwrt-kirkwood-uImage and openwrt-kirkwood-rd88f6281-squashfs.uimg.
+
+Follow below steps to install firmware in to device's NAND flash.
+(Remember to use apropriate network parameters)
+
+1. Prepare uBoot environment (In uBoot):
+	>>resetenv; reset
+	>>setenv mainlineLinux yes; save; reset
+	>>setenv ipaddr 192.168.2.200; setenv serverip 192.168.1.2
+	>>setenv loadaddr 0x2000000; setenv console 'console=ttyS0,115200 panic=30'
+	>>setenv bootargs_root 'ubi.mtd=2 root=ubi0:rootfs rootfstype=ubifs rw'
+	>>setenv image_name openwrt-kirkwood-uImage
+	>>setenv root_name openwrt-kirkwood-rd88f6281-squashfs.uimg
+	>>setenv update_image 'tftpboot $(loadaddr) $(image_name); nand erase 0x100000 0x200000; nand write $(loadaddr) 0x100000 0x200000'
+	>>setenv load_firmware 'tftpboot $(loadaddr) $(root_name); setenv bootargs $(console) root=/dev/ram0 rw; bootm $(loadaddr)'
+	>>setenv bootcmd 'setenv bootargs $(console) $(bootargs_root); nand read $(loadaddr) 0x100000 0x200000; bootm $(loadaddr)'
+	>>save
+
+2. Load initial firmare (In uBoot):
+	>>run update_image
+	>>run load_firmware
+
+3. Install firmware into NAND (In Linux):
+	$ install2ubifs
+	$ reboot
Index: openwrt/package/uboot-kirkwood/files/board/Marvell/iconnect/iconnect.c
===================================================================
--- openwrt/package/uboot-kirkwood/files/board/Marvell/iconnect/iconnect.c	(revision 29704)
+++ openwrt/package/uboot-kirkwood/files/board/Marvell/iconnect/iconnect.c	(working copy)
@@ -100,7 +100,7 @@
 	/*
 	 * arch number of board
 	 */
-	gd->bd->bi_arch_number = MACH_TYPE_ICONNECT;
+	gd->bd->bi_arch_number = 1682;//MACH_TYPE_ICONNECT;
 
 	/* adress of boot parameters */
 	gd->bd->bi_boot_params = kw_sdram_bar(0) + 0x100;
