--- transmission-2.60/libtransmission/version.h	2012-07-06 11:27:22.331652591 +1000
+++ transmission-2.33/libtransmission/version.h	2011-07-21 09:39:34.135997092 +1000
@@ -1,10 +1,10 @@
-#define PEERID_PREFIX             "-TR2600-"
-#define USERAGENT_PREFIX          "2.60"
-#define SVN_REVISION              "13375"
-#define SVN_REVISION_NUM          13375
-#define SHORT_VERSION_STRING      "2.60"
-#define LONG_VERSION_STRING       "2.60 (13375)"
-#define VERSION_STRING_INFOPLIST  2.60
+#define PEERID_PREFIX             "-TR2330-"
+#define USERAGENT_PREFIX          "2.33"
+#define SVN_REVISION              "12565"
+#define SVN_REVISION_NUM          12565
+#define SHORT_VERSION_STRING      "2.33"
+#define LONG_VERSION_STRING       "2.33 (12565)"
+#define VERSION_STRING_INFOPLIST  2.33
 #define MAJOR_VERSION             2
-#define MINOR_VERSION             60
+#define MINOR_VERSION             33
 #define TR_STABLE_RELEASE         1
