diff --git a/target/linux/kirkwood/Makefile b/target/linux/kirkwood/Makefile
index 426ffd2..81afaf5 100644
--- a/target/linux/kirkwood/Makefile
+++ b/target/linux/kirkwood/Makefile
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 ARCH:=arm
 BOARD:=kirkwood
 BOARDNAME:=Marvell Kirkwood
-FEATURES:=targz usb jffs2
+FEATURES:=targz usb jffs2 ubifs
 MAINTAINER:=Imre Kaloz <kaloz@openwrt.org>
 
 LINUX_VERSION:=3.2.9
diff --git a/target/linux/kirkwood/image/Makefile b/target/linux/kirkwood/image/Makefile
index c93be67..9e60b88 100644
--- a/target/linux/kirkwood/image/Makefile
+++ b/target/linux/kirkwood/image/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2009-2010 OpenWrt.org
+# Copyright (C) 2009-2012 OpenWrt.org
 #
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
@@ -9,6 +9,19 @@ include $(INCLUDE_DIR)/image.mk
 
 JFFS2OPTS += --little-endian --pagesize=0x800 --no-cleanmarkers --pad
 
+ifeq ($(PROFILE),Sheevaplug)
+UBIFS_OPTS += -m 2048 -e 126KiB -c 8192
+UBINIZE_OPTS += -m 2048 -p 128KiB -s 512
+endif
+ifeq ($(PROFILE),Dockstar)
+UBIFS_OPTS += -m 2048 -e 126KiB -c 4096
+UBINIZE_OPTS += -m 2048 -p 128KiB -s 512
+endif
+ifeq ($(PROFILE),Iconnect)
+UBIFS_OPTS += -m 2048 -e 126KiB -c 8192
+UBINIZE_OPTS += -m 2048 -p 128KiB -s 512
+endif
+
 define Image/Prepare
 	cp $(LINUX_DIR)/arch/arm/boot/uImage $(KDIR)/uImage
 endef
diff --git a/target/linux/kirkwood/image/ubinize.cfg b/target/linux/kirkwood/image/ubinize.cfg
new file mode 100644
index 0000000..49d55b9
--- /dev/null
+++ b/target/linux/kirkwood/image/ubinize.cfg
@@ -0,0 +1,14 @@
+[rootfs]
+# Volume mode (other option is static)
+mode=ubi
+# Source image
+image=root.ubifs
+# Volume ID in UBI image
+vol_id=0
+# Allow for dynamic resize
+vol_type=dynamic
+# Volume name
+vol_name=rootfs
+# Autoresize volume at first mount
+vol_flags=autoresize
+
