diff --git a/target/linux/kirkwood/Makefile b/target/linux/kirkwood/Makefile
index 06b6fc7..426ffd2 100644
--- a/target/linux/kirkwood/Makefile
+++ b/target/linux/kirkwood/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2009-2011 OpenWrt.org
+# Copyright (C) 2009-2012 OpenWrt.org
 #
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
@@ -12,7 +12,7 @@ BOARDNAME:=Marvell Kirkwood
 FEATURES:=targz usb jffs2
 MAINTAINER:=Imre Kaloz <kaloz@openwrt.org>
 
-LINUX_VERSION:=2.6.37.6
+LINUX_VERSION:=3.2.9
 
 include $(INCLUDE_DIR)/target.mk
 
diff --git a/target/linux/kirkwood/config-default b/target/linux/kirkwood/config-default
index d37d8ab..a0d7e94 100644
--- a/target/linux/kirkwood/config-default
+++ b/target/linux/kirkwood/config-default
@@ -9,6 +9,7 @@ CONFIG_ARCH_SUSPEND_POSSIBLE=y
 # CONFIG_ARCH_USES_GETTIMEOFFSET is not set
 CONFIG_ARM=y
 CONFIG_ARM_L1_CACHE_SHIFT=5
+CONFIG_ARM_PATCH_PHYS_VIRT=y
 # CONFIG_ARM_THUMB is not set
 # CONFIG_ARPD is not set
 CONFIG_BITREVERSE=y
diff --git a/target/linux/kirkwood/patches/200-iconnect.patch b/target/linux/kirkwood/patches/200-iconnect.patch
index f8baae6..30693a8 100644
--- a/target/linux/kirkwood/patches/200-iconnect.patch
+++ b/target/linux/kirkwood/patches/200-iconnect.patch
@@ -218,7 +218,7 @@
 +
 +
 +MACHINE_START(ICONNECT, "Iomega iConnect Wireless")
-+	.boot_params	= 0x00000100,
++	.atag_offset	= 0x100,
 +	.init_machine	= iconnect_init,
 +	.map_io		= kirkwood_map_io,
 +	.init_irq	= kirkwood_init_irq,
